File: ServerView.java
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;

public class serverView { 


          public static void main(String[] args) throws IOException {
         
             int port = 9998; //Define the port on which server listen to clients
             ServerSocket serverSocket = new ServerSocket(port);
             System.out.println("Server is running and listening to port" + port);
             //Accept client connection 
             Socket clientSocket = serverSocket.accept(); //Server is looking for a client to connect
             BufferedReader inFromClient = new BufferedReader(new InputStreamReader(clientSocket.getInputStream())); // Create a reader to receive maessages from the clients
             DataOutputStream outToClient = new DataOutputStream(clientSocket.getOutputStream()); // Create a stream to send responses to the clients

             // Create a loop to continuously receive message from clients 
             while (true) { 
                 String clientMessage = inFromClient.readLine(); // Read the clients' message 

                 // If the client sends "Quit", break the loop and cut the connection 
                 if (clientMessage.equalsIgnoreCase("Quit")) { // check if client wants to quit  

                    System.out.println("Client has exited");
                    break; // Break the loop and stop the server
                 } // End First if 

                 // Split message and checksum in format (meesage:checksum)
                 String[] messageParts = clientMessage.split(":");

                 if (messageParts.length != 2) { // Check of message format validity
                    outToClient.writeBytes(("Error: Invalid message format\n")); // Send error if the format is wrong
                    continue; // Skip this iteration and wait for the message


                 } // End second if 

                 String message = messageParts[0];
                 String receivedChecksum = messageParts[1]; // Extract the received checkusm
                 System.out.println("Received Message: " + message + "With checkum" + receivedChecksum); // Print out received message from the client

                 // Caluclate checksum of the received message
                 String calculateChecksum = calculateChecksum(message);

                 // Compare calculated checksum with received checksum

                 if (!calculateChecksum.equals(receivedChecksum))  { //if checksum don't match
                     outToClient.writeBytes("Error: The Received Message is not correct\n"); // Send error message to the client
                  } else {
                     outToClient.writeBytes("Confirmation: Message received correctly \n");

                  }

          

               } // End the loop 

               // Close the connection and server socket

               clientSocket.close();
               serverSocket.close();


          } // End main 

             // checksum calculating


          private static String calculateChecksum(String message) {
               int checksum = 0;
                

               //loop through the message 16-bits at a time
               for (int i = 0; i < message.length(); i += 2) {
                  int word = 0;

                  // Get the first charachter (8 bits)
                  word = message.charAt(i) << 8; // Shift the first charcter 8 bits to the left (upper byte)

                  // Get the second charcter if it exists
                  if (i + 1 < message.length()) {
                     word += message.charAt(i + 1); // Add the second charcter (lower byte)

                  }

                  // Add the 16-bit word to the checksum
                  checksum += word;

                  // Handle carry (wrap around) if the sum exceeds 16 bits

                  if ((checksum & 0xFFFF0000) > 0) { // If ther's a carry beyond 16 bits 

                  checksum &= 0xFFFF; // Keep only the lower 16 bits
                  checksum++; // Add the carry

                  }
                                     
               } // End loop 

               // One's complement
               checksum = ~checksum;

               // Print the checksum as a binary string and hexdecimal value
               System.out.println("Checksum (binary):" + String.format("%16s", Integer.toBinaryString(checksum)).replace(' ', '0'));
               System.out.println("Checksum (hexdecimal): " + Integer.toBinaryString(checksum).toUpperCase());


               // Return the checksum as a hexdecimal string
               return  Integer.toHexString(checksum).toUpperCase();


 
            }     
          
     

             
             

              


} // End the class

        
    

    

File: serverView.py
import socket
import random

# Function to calculate 16-bit one's complement checksum
def calculate_checksum(message):
    checksum = 0
    
    for i in range(0, len(message), 2):# Process the message in chunks of 2 bytes (16 bits)
        # Create a 16-bit word from two characters
        word = ord(message[i]) << 8  # Get first character (upper byte) ,
        #The ord() function returns the number of unicode code of a specified character.
        if i + 1 < len(message):  # Check if there's a second character
            word += ord(message[i + 1])  # Add second character (lower byte)
        checksum += word  # Sum the 16-bit word
        if checksum & 0x10000:  # Handles any carry that occurs if the sum exceeds 16 bits.
            checksum = (checksum & 0xFFFF) + 1  # Handle carry (wrap around)
    checksum = ~checksum & 0xFFFF  # One's complement (invert all bits)
    return format(checksum, '016b')  # Return as a 16-bit binary string

def start_server():
    server_address = '192.168.100.160'  #TEST Server IP address
    port = 8080  # the port of the server 

    # Create the server socket
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind((server_address, port))  # Bind to the address and port
    server_socket.listen(1)  #  The server will allow one connection to wait in the queue(for simple server)
    
    print(f"The Server is Available and wait for Client on port {port}")

    # if you want the server is ALWAYS ON HOST use here infnite while true without break
    # this will make the server run even the client disconnect :) 
    # but we dont have the best resourcses for infnite loops :(  
    
    # Accept client connection
    client_socket, client_address = server_socket.accept()
    print(f"Client connected from {client_address}")

    while True:
        client_message = client_socket.recv(1024).decode('utf-8')  # Receive message

        if client_message.lower() == 'quit':  # Check if the client wants to quit
            print("Client has exited.")
            break

        # Split message into data and checksum
        message_parts = client_message.split(':')
        if len(message_parts) != 2:  # Ensure message is in the correct format
            client_socket.sendall("Error: Invalid message format\n".encode('utf-8'))
            continue

        message= message_parts[0]
        received_checksum =message_parts[1]
        print(f"Received Message: {message} with Checksum: {received_checksum}")

        # Calculate checksum for the received message
        calculated_checksum = calculate_checksum(message)







        # Check if calculated checksum matches the received checksum
        if calculated_checksum != received_checksum:
            client_socket.sendall(f"Error: The Received Message is not correct with server checksum: {calculated_checksum} client checksum: {received_checksum}\n".encode('utf-8'))
        else:
            client_socket.sendall(f"Acknowledged: Message received correctly checksum: {calculated_checksum} client checksum: {received_checksum}\n".encode('utf-8'))
#the\n helps ensure that when this error message is displayed on the client side, it’s printed correctly.
    
    client_socket.close()  # Close client connection
    server_socket.close()  # Close server

if __name__ == "__main__":
    start_server()  # Start the server
